sensors: add RCInput message

peterbarker opened the pull request:
WIP; discuss before merge

David-OConnor commented:
I would remove RSSI; it's better to have it in a separate packet as:

- RSSI is one picture of link stats and it's not as useful as say, link quality (% of recent packets received). 
- There are many other things that should accompany RSSI, like link quality, antenna selection, RSSI and link stats for a second antenna, Tx power etc.
- It's generally sent at a significantly lower rate packet than control channel data.

Small point as you could ignore the field, but IMO it complicates things.

AlexKlimaj commented:
Yeah a lower rate status packet would be fine. But it also is such a small difference in overall packet size having it in this message.

David-OConnor commented:
Concur on the added bytes not being a big deal.

dagar commented on dronecan/sensors/rc/20400.RCInput.uavcan:
```diff
@@ -0,0 +1,6 @@
+uint16 status
+
+uint8 rssi
+uint12[<=24] rcin
```
description

dagar commented on dronecan/sensors/rc/20400.RCInput.uavcan:
```diff
@@ -0,0 +1,6 @@
+uint8 STATUS_FAILSAFE = 2 # receiver has lost contact with transmitter
+uint16 status
+
+uint8 rssi
```
percentage? How to handle the difference between not provided and zero?

tridge replied:
there is a status bit for provided

peterbarker commented:
On Sat, 20 May 2023, Daniel Agar wrote:

> 
> @dagar commented on this pull request.
> 
> ____________________________________________________________________________
> 
> In dronecan/sensors/rc/20400.RCInput.uavcan:
> 
> > @@ -0,0 +1,6 @@
> +uint8 STATUS_RSSI_VALID = 1 # RSSI field is valid
> +uint8 STATUS_FAILSAFE = 2 # receiver has lost contact with transmitter
> +uint16 status
> +
> +uint8 rssi
> 
> percentage? How to handle the difference between not provided and zero?

First bit in the status field for rssi validity.

I've presumed percentage for RSSI but I'm open to suggestions.

dagar commented:
> I've presumed percentage for RSSI but I'm open to suggestions.

I'm not sure, we can't really win here because there are a lot different options. In newer systems like CRSF/ELRS it's actually in dBm and separately there's a link quality percentage (% of uncorrupted packets I believe).

Either way we should specify in a comment for the field so there's no ambiguity.

Another idea to keep in mind if the range of options becomes clear is that unions are possible in UAVCAN.

David-OConnor commented:
What is the intent behind 2-6 bit channels? This is concretely answered in FC implementations and doesn't necessarily affect this PR directly. Is the intent to pack into the 12-bit channels, or assign one per channel. Note that if it's the latter, multiple 2 and 3 position switches will cause a large bandwidth use. (Eg ~200 bytes sent at a rate of 1Khz depending on how many channels and thebroadcast rate)

hendjoshsr71 commented on dronecan/sensors/rc/20400.RCInput.uavcan:
```diff
@@ -0,0 +1,6 @@
+uint16 status
+
+uint8 rssi
+uint12[<=24] rcin
```
Thinking about the future we already support 32 servos in AP. And  just because right now most receivers only receive a limited number of channels doesn't mean they always will...

```suggestion
uint12[<=24] rcin
uint8 index # Index of the first rc_channel in the array, index 0 is rc_channel at array indices 0 - 23, index 24 is rc_channel at array indices 24 - 47, etc.
```

peterbarker replied:
> Thinking about the future we already support 32 servos in AP. And just because right now most receivers only receive a limited number of channels doesn't mean they always will...

Overkill :-)  Most people aren't using 16 channels, let alone 24 :-)

David-OConnor commented:
Re the comments about keeping it simple and sticking to existing use cases: I'm fine with that *if the team is open to merging updated messages or breaking old messages later*.

peterbarker commented:
> Concur on the added bytes not being a big deal. My leaning against it is to prevent the question "if rssi here is 0, does that mean there is no rssi data? Or does it mean reference a link stats packet? What if the two disagree?"
> 
> You could answer those questions, but given a choice, I'd not ask them.

RSSI may not be a valid field; that's what the status bit means.  What 0 means when num_channels is non-zero is undefined.  My current implementation won't send any DroneCAN packets when there's no RC input to the AP_Periph node, so it's not a problem there.

peterbarker commented:
> > I've presumed percentage for RSSI but I'm open to suggestions.
> 
> I'm not sure, we can't really win here because there are a lot different options. In newer systems like CRSF/ELRS it's actually in dBm and separately there's a link quality percentage (% of uncorrupted packets I believe).
> 
> Either way we should specify in a comment for the field so there's no ambiguity.
> 
> Another idea to keep in mind if the range of options becomes clear is that unions are possible in UAVCAN.

I've added a comment in - 0-255

peterbarker commented:
I've sprinkled some comments in, but if I don't hear some serious screaming we're probably going to call this done.

I think a link status packet in the future would be a good idea - but RSSI is sufficient for most purposes.

David-OConnor commented:
Hi! Could someone please post the base CRC? Thank you!

peterbarker commented:
I've moved this to a "more correct" ID.

I've increased the number of channels to 32 (@hendjoshsr71 )

I've renamed "rssi" to "quality" (@dagar )
